JVM: New single-pass implementation for jvm-abi-gen

sfs opened the pull request:
The `jvm-abi-gen` plugin currently runs the Kotlin compiler backend twice - once to generate ABI classes and then again to produce the compiler output. This is a bit slower and more conservative than it has to be.

This PR contains a new single-pass implementation of jvm-abi-gen and also adds more checks to ensure that the implementation is correct. The previous implementation remains usable with a command line flag (`useLegacyAbiGen`), in case we have users who rely on the current implementation.

The main change in the compiler is a new flag in the Kotlin metadata which marks classes that are implicitly part of the public ABI of a Kotlin library. This is the case for:
- Anonymous objects in the scope of an externally visible inline function (this only excludes private and private-to-this, since `internal` functions may be visible through `-Xfriend-modules`).
- WhenMapping classes used from externally visible inline functions.
- SAM wrappers used in inline functions<sup>[1](#samfootnote)</sup>.

I also added an assertion to the `AnonymousObjectTransformer` to check that all recreated anonymous objects from other modules are marked as public, in case we add another special case in the future.

The `jvm-abi-gen` plugin then uses this flag to determine which classes are part of the public ABI and marks them in an initial `ClassBuilderInterceptor`. The actual ABI class/jar generation happens as part of a new `ClassFileFactoryFinalizerExtension` extension point. This is necessary in order to handle *all* generated files, otherwise we might miss `kotlin_module` files or  files generated by other compiler plugins.

---

<a name="samfootnote">1</a>: Ideally, we would only create inline SAM wrappers (`...$sam$i$...`) if the wrapper is used in the scope of an externally visible inline function. However, this would change the names of generated classes and is a bit out of scope for this PR.

sfs commented:
Rebased on master

sfs commented:
I've changed the codegen to preserve more metadata, strip out synthetic accessors, and sort declarations. I've also added some more tests, although there is still room for improvement here.

udalov commented on compiler/backend/src/org/jetbrains/kotlin/codegen/when/MappingClassesForWhenByEnumCodegen.java:
```diff
@@ -57,7 +57,15 @@ public void generate(@NotNull List<WhenByEnumsMapping> mappings, @NotNull Type m
         generateInitialization(cb, mappings);
 
-        WriteAnnotationUtilKt.writeSyntheticClassMetadata(cb, state);
+        boolean publicAbi = false;
+        for (WhenByEnumsMapping mapping : mappings) {
+            if (mapping.isPublicAbi()) {
+                publicAbi = true;
+                break;
+            }
+        }
```
```suggestion
        boolean publicAbi = mappings.stream().anyMatch(WhenByEnumsMapping::isPublicAbi);
```

sfs replied:
That's neat, thanks!

udalov commented on compiler/backend/src/org/jetbrains/kotlin/codegen/SamWrapperCodegen.java:
```diff
@@ -133,7 +133,7 @@ public Type genWrapper(
         cv.visitSource(file.getName(), null);
 
-        WriteAnnotationUtilKt.writeSyntheticClassMetadata(cv, state);
+        WriteAnnotationUtilKt.writeSyntheticClassMetadata(cv, state, isInsideInline);
```
Shouldn't we also check here that it's inside a non-private inline function?

sfs replied:
Sure, I can add that. Alternatively, I was thinking that going forward we could only generate inline SAM wrappers from within public inline functions. Since these classes are regenerated at call sites this wouldn't be a breaking change either and might improve SAM wrapper caching. What do you think?

udalov replied:
So you mean changing the logic which computes `isInsideInline` in `SamWrapperClasses.getSamWrapperClass` to check only if it's inside public inline? Yes, sounds good too!

sfs replied:
Great, done! I've also added tests for this, and got rid of the odd behavior where we would generate an inline sam wrapper when it was used from a lambda argument to an inline function. I'm really confused as to why we did that.

sfs replied:
I somehow didn't run all tests for this change and missed that this broke `kt24825`. Incidentally, that test is probably the reason for the JVM BE behavior around SAM wrappers in inline lambdas.

Unresolving this thread until I can figure out what went wrong there.

sfs replied:
The bug was in the inliner, which identified non-inline SAM wrappers as anonymous objects to be regenerated and would then remove the original wrapper.

Since this only affects non-public inline functions and hence only inline calls from within the same model I just fixed this in the inliner.

udalov commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiMetadataProcessor.kt:
```diff
@@ -0,0 +1,177 @@
+        override fun visit(name: String, value: Any?) {
+            when (name) {
+                "k" -> kind = value as Int
+                "extraInt" -> extraInt = value as Int
```
I think it should be `"xi"`

sfs replied:
Right... I switched to using the constants from `JvmAnnotationNames`.

udalov commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiMetadataProcessor.kt:
```diff
@@ -0,0 +1,177 @@
+            return super.visitTypeAlias(flags, name)
+        return null
+    }
+}
```
Shouldn't metadata of local delegated properties declared in top level functions be removed here, similarly to how it's being removed for classes in `AbiKmClassVisitor`?

sfs replied:
Right. It's just that `JvmPackageExtensionVisitor` does not seem to have any methods for local delegated properties. Are we even recording top-level local delegated properties in the metadata?

udalov replied:
`JvmPackageExtensionVisitor` extends from `JvmDeclarationContainerExtensionVisitor` (like `JvmClassExtensionVisitor`), which declares `visitLocalDelegatedProperty`, so exactly the same approach with overriding `visitExtensions` should work

sfs replied:
Wow, I have no idea how I kept missing this. Thanks for pointing it out. It's fixed now.

udalov commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiMetadataProcessor.kt:
```diff
@@ -0,0 +1,177 @@
+ * Class metadata adapter which removes private functions, properties, type aliases,
+ * and local delegated properties.
+ */
+private class AbiKmClassVisitor(delegate: KmClassVisitor) : KmClassVisitor(delegate) {
```
Shouldn't we also remove private nested classes, and private constructors?

sfs replied:
Yes, absolutely. Fixed and also removed private Companion objects.

I guess we should not remove private sealed subclasses due to exhaustiveness checking, though?

udalov replied:
Oh. I wonder if removing the nested classes and keeping the references in sealed subclasses breaks the exhaustiveness checking anyway, because the compiler can't load the nested class correctly and thus may skip loading it altogether (or worse, crash with some exception).

sfs replied:
That part is not an issue right now, since we only strip nested anonymous classes.

This is necessary for Java, since the classes will still be package visible, but I wonder if we should strip more classes after all. In principle this could break Java code, though, so I'm not completely convinced. What do you think?

udalov replied:
Sorry for the delay. I'm probably missing something. My initial comment was about private nested classes, and we now do strip them, after a new change you've added.

However, now I worry that that might have been risky, because the Kotlin compiler treats the "nested classes" field as the only source of truth. A class named `A$B` _possibly_ won't be treated as a nested class named `B` inside `A`, unless `A` has `B` in its nested classes in the metadata. Whether that would lead to any noticeable problems is an open question though.

In case of private sealed subclasses which are nested, the metadata becomes _sort of_ inconsistent now -- the "sealed subclasses" field mentions some classes which are clearly nested in their outer class, yet that outer class knows nothing about them, because they are absent in its "nested classes" field in the metadata. Again, I'm not sure that it would cause any problems.

From the data organization point of view, all nested classes should be declared in the "nested classes" field. Any other data, including "companion object" and "sealed subclasses", does not _declare_ new classes, but only references the existing ones declared in "nested classes".

sfs replied:
On reflection, the current code seems to be completely useless. Right now we only strip (some) local anonymous classes. In particular, none of the classes that we remove would end up in the Kotlin metadata in the first place. We might want to revisit this later, since there are definitely more classes that we could strip, but for now I've removed the code in question.

udalov commented:
@max-kammerer Please take a look (at least) at the first commit

max-kammerer commented on compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ClassCodegen.kt:
```diff
@@ -208,7 +208,18 @@ class ClassCodegen private constructor(
+        //    can be reused in different functions and are thus generated in the enclosing top-level
+        //    class instead of inside of an inline function.
+        // 3) WhenMapping classes used from inline functions. These are collected in
+        //    `JvmBackendContext.publicAbiSymbols` in `MappedEnumWhenLowering`.
```
In future we can support lambda/sam/function reference generation via invokedynamic. Would it affect current scheme in any sense? Should we write any flags for bootstrap methods?

sfs replied:
The bootstrap method should be unaffected, since it probably won't contain any anonymous objects. The only potential problem that I can think of is that the invoke method will be lowered into a private method in the enclosing class. We should ensure that we add a reference to the original `IrFunction` via `attributeOwnerId` so that `isInPublicInlineScope` keeps working.

max-kammerer commented on compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ClassCodegen.kt:
```diff
@@ -208,7 +208,18 @@ class ClassCodegen private constructor(
+        // 3) WhenMapping classes used from inline functions. These are collected in
+        //    `JvmBackendContext.publicAbiSymbols` in `MappedEnumWhenLowering`.
+        val isPublicAbi = irClass.symbol in context.publicAbiSymbols || type.isInlineSamWrapper ||
+                type.isAnonymousClass && irClass.isInPublicInlineScope
```
`type.isAnonymousClass` could `irClass` be used instead of type here?

sfs replied:
Using `irClass.isAnonymousObject` ends up missing some cases here (e.g. in `testInnerMadness`). The logic needs to match the `AnonymousObjectTransformer` and this was the easiest way. However, we can look at `irClass` to identify inline sam wrappers and I've changed the code accordingly.

max-kammerer replied:
@udalov we need to think which way is less fragile in long perspective. Current approach is looking OK

sfs commented:
Rebased on master and fixed SAM wrapper generation in the old backend.

udalov commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiClassBuilderInterceptor.kt:
```diff
@@ -0,0 +1,157 @@
+ * two corner cases to consider.
+ *
+ * First, if we have an inline function which is not callable from Java (i.e.,
+ * due to reified type parameters), we generate a private method, which nevertheless
```
This sentence doesn't seem correct... We're currently generating public synthetic methods for public inline+reified functions. We did generate them as private, but that was before 1.3, I think

sfs commented:
Sorry for the slow turnaround on this PR. I've spent some time to testing the implementation and fixed a few issues (see new commits - apart from that there are only some small changes in the existing code).

sfs commented:
Rebased on master and changed a command line option name to avoid breaking existing users of `jvm-abi-gen`.

sfs commented:
Rebased on master after some inliner changes.

max-kammerer commented on compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/ir/IrUtils.kt:
```diff
@@ -417,3 +419,25 @@ inline fun IrElement.hasChild(crossinline block: (IrElement) -> Boolean): Boolea
+    }
+
+// True for declarations which are in the scope of an externally visible inline function.
+val IrDeclaration.isInPublicInlineScope: Boolean
```
What about `internal` case?

sfs replied:
I don't think we can safely ignore `internal` declarations, since they need to be visible when using `-Xfriend-paths`.

max-kammerer replied:
Yes, indeed. And sorry for pause in review

max-kammerer commented on compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/ir/IrUtils.kt:
```diff
@@ -417,3 +419,25 @@ inline fun IrElement.hasChild(crossinline block: (IrElement) -> Boolean): Boolea
+// Declarations in the scope of an externally visible inline function are implicitly part of the
+// public ABI of a Kotlin module. This function returns the visibility of a containing inline function
+// (determined *before* lowering), or null if the given declaration is not in the scope of an inline function.
+val IrDeclaration.inlineScopeVisibility: DescriptorVisibility?
```
Could you also add (abi?) test for next case:
```
var result = "fail"

inline fun foo(crossinline s: () -> String) {

    object {
        private inline fun test(crossinline z: () -> String) {
            {
                result = z()
            }()
        }

        fun foo() {
            test { s() }
        }
    }.foo()
}

fun box() {
    foo {
        "OK"
    }
}

fun main() {
    box()
    println(result)
}
```

max-kammerer replied:
and without inner object regeneration (I've also updated snippet above a bit):
```
var result = "fail"

inline fun foo(crossinline s: () -> String) {

    object {
        private inline fun test(crossinline z: () -> String) {
            result = object {
                fun run(): String {
                    return "O"
                }
            }.run() + z()
        }

        fun foo() {
            test { s() }
        }
    }.foo()
}

fun box() {
    foo {
        "K"
    }
}

fun main() {
    box()
    println(result)
}
```

sfs replied:
I'm consistently amazed by how quickly you can come up with these convoluted examples.

This was not handled correctly, as I mistakenly assumed that anonymous objects inside of private inline functions could not escape. Your second example shows that this is not the case. I've fixed this problem and added the tests (in a new commit to make it easier to review).

On a side note, the first test is also broken in the legacy jvm-abi-gen implementation.

max-kammerer replied:
> I'm consistently amazed by how quickly you can come up with these convoluted examples.

Thank you, nice to hear!
But still some cases could be missed anyway, and I've forgot about next ones
```
class A {
    private class B {
        inline fun test(crossinline s: () -> Unit) {
            val z = 1;
            object {
                fun run() {
                    s()
                }
            }.run()
        }
    }
}

private class C {
    inline fun test(crossinline s: () -> Unit) {
        val z = 1;
        object {
            fun run() {
                s()
            }
        }.run()
    }
}
```

max-kammerer commented on generators/tests/org/jetbrains/kotlin/generators/tests/GenerateTests.kt:
```diff
@@ -171,15 +169,45 @@ fun main(args: Array<String>) {
+            }
+        }
+
+        testGroup(
```
Why these tests are not targeted to IR?

sfs replied:
Because the old implementation of jvm-abi-gen does not really support the IR backend. It is always running the old backend as part of `JvmAbiAnalysisHandlerExtension` ([here](https://github.com/JetBrains/kotlin/blob/a54e758189c2048c5d5550936fad3d8afa5c1703/plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiAnalysisHandlerExtension.kt#L54)).

...and this just reminded me that the tests need to pass `useOldBackend` instead of `useIR` now... that's also fixed.

max-kammerer commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiComponentRegistrar.kt:
```diff
@@ -14,13 +16,23 @@ import java.io.File
-            val dir = File(it)
-            configuration.put(JVMConfigurationKeys.OUTPUT_DIRECTORY, dir)
+        val outputPath = configuration.getNotNull(JvmAbiConfigurationKeys.OUTPUT_PATH)
+        if (configuration.get(JvmAbiConfigurationKeys.LEGACY_ABI_GEN, false)) {
```
Do I understand correctly that new scheme would be default one after merging in 1.5.30?
I would prefer to announce it as non-default option in 1.5.30 if it's ok and make default in 1.6 // cc @udalov

udalov replied:
It seems that the plugin won't make it to 1.5.30, we hope to merge it into 1.6-M1.

I think it's fine to change the behavior right away, since the new scheme is objectively better and that way we'll gather feedback quicker. Also, there's not a lot of users of the plugin, so I don't think it'll break a lot.

sfs replied:
There are also known issues with the old scheme (e.g., KT-40133 and KT-40340), which are fixed with the new implementation.

max-kammerer commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiMetadataProcessor.kt:
```diff
@@ -0,0 +1,194 @@
+        val newHeader = when (val metadata = KotlinClassMetadata.read(header)) {
+            is KotlinClassMetadata.Class -> {
+                val writer = KotlinClassMetadata.Class.Writer()
+                metadata.accept(AbiKmClassVisitor(writer))
```
What about private classes (top level,  nested)?

max-kammerer replied:
How they skipped, processed?

sfs replied:
Right now they are kept in the metadata. The reasoning is that since we don't strip private classes from the ABI (as they are still visible from Java) it's probably for the best to keep them in the metadata as well. Otherwise they wouldn't be recognized as nested classes at all, which could lead to further problems.

max-kammerer replied:
Is it any test for that?

max-kammerer replied:
maybe with Java client

sfs replied:
Good point, I've added tests for private (annotation) classes in ba5417d.

max-kammerer commented:
Do I understand correctly that tests are checking only compilation against ABI classes. So ABI could potentially contain explicit information?

max-kammerer commented on compiler/testData/codegen/bytecodeText/sam/samWrapperInInlineLambda.kt:
```diff
@@ -16,4 +16,5 @@ fun box(): String {
 
-// 1 INVOKESPECIAL TestKt\$sam\$i\$java_lang_Runnable\$0.<init> \(Lkotlin/jvm/functions/Function0;\)V
\ No newline at end of file
+// 1 INVOKESPECIAL TestKt\$sam\$java_lang_Runnable\$0.<init> \(Lkotlin/jvm/functions/Function0;\)V
```
@sfs Could you also same test in case when box is inline

sfs replied:
Done (`compiler/testData/codegen/bytecodeText/sam/samWrapperInInlineLambda2.kt`)

max-kammerer commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiClassBuilderInterceptor.kt:
```diff
@@ -0,0 +1,162 @@
+
+        override fun defineClass(origin: PsiElement?, version: Int, access: Int, name: String, signature: String?, superName: String, interfaces: Array<out String>) {
+            // Always keep annotation classes
+            if (access and Opcodes.ACC_ANNOTATION != 0) {
```
Is there any test for private annotations?

sfs replied:
To be honest, this was added to fix a problem that occurred in our internal codebase and I don't remember what exactly the issue was.

In general we need to keep annotation classes for Java compatibility, but I don't remember which aspect of an annotation class is in danger of being stripped at all. I'll try to reproduce the failure once I can run said internal codebase with a more up to date version of the Kotlin compiler again.

Meanwhile I've added a test in `plugins/jvm-abi-gen/testData/content/annotation` so that we have at least some test coverage for annotation classes.

max-kammerer commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiMetadataProcessor.kt:
```diff
@@ -0,0 +1,194 @@
+ */
+fun abiMetadataProcessor(annotationVisitor: AnnotationVisitor): AnnotationVisitor =
+    kotlinClassHeaderVisitor { header ->
+        // kotlinx-metadata only supports writing Kotlin metadata of version >= 1.4, so we need to
```
@sfs could you clarify a bit about scenario when this is necessery? I'm not familar enough with this tooling. Could it be used as separate tool? 
@udalov How good to force metadata upgrade in such way?

udalov replied:
I don't see any way around it. Supporting older metadata versions in `kotlinx-metadata-jvm` would be a bit complicated. Probably it should be explicitly documented in some kind of docs on `jvm-abi-gen` though

udalov replied:
Yeah, since this PR has started, we removed support of language/metadata version 1.3 from the compiler, so this is no longer a concern.

max-kammerer commented:
@sfs Abi tests are failed on TC with
```
java.lang.NoSuchMethodError: com.intellij.openapi.util.text.StringUtil.trimLeading(Ljava/lang/String;)Ljava/lang/String;
  at org.jetbrains.kotlin.test.InTextDirectivesUtils.cleanDirectivesFromComments(InTextDirectivesUtils.java:172)
  at org.jetbrains.kotlin.test.InTextDirectivesUtils.findLinesWithPrefixesRemoved(InTextDirectivesUtils.java:118)
  at org.jetbrains.kotlin.test.InTextDirectivesUtils.findLinesWithPrefixesRemoved(InTextDirectivesUtils.java:109)
  at org.jetbrains.kotlin.test.InTextDirectivesUtils.findListWithPrefixes(InTextDirectivesUtils.java:65)
  at org.jetbrains.kotlin.test.InTextDirectivesUtils.isIgnoredTarget(InTextDirectivesUtils.java:245)
  at org.jetbrains.kotlin.test.KotlinTestUtils.lambda$testWithCustomIgnoreDirective$5(KotlinTestUtils.java:567)
  at org.jetbrains.kotlin.test.KotlinTestUtils.runTestImpl(KotlinTestUtils.java:547)
```

sfs commented:
Rebased on master and fixed some issues; see comments above.

As for the build failures on TC, as far as I can tell these tests should never have worked on TC. There was a configuration problem in `plugins/jvm-abi-gen/build.gradle.kts` which listed the intellij dependency after the compiler dependencies. I've changed this in 1c039d2. Could you please check that this fixes the build issues @max-kammerer ?

sfs commented:
> Do I understand correctly that tests are checking only compilation against ABI classes. So ABI could potentially contain explicit information?

Yes and in fact the "ABI" is currently an over approximation. For example, we are not yet stripping unused line numbers from the Kotlin debug info. I'm leaving this for future work, same as a more aggressive ABI generation mode which would also strip private classes (hence not be a drop-in replacement as far as Java is concerned, but probably a better default for most users).

sfs commented:
@max-kammerer This was another configuration issue, this time with the test generator. I changed the dependency on the `:compiler:incremental-compilation-impl` tests in `:plugins:jvm-abi-gen` from the deprecated `testCompile` to `testImplementation`. The difference between the two is that the latter is not transitive and apparently the test generator was only depending on `:compiler:incremental-compilation-impl` through its dependency on `:plugins:jvm-abi-gen`. My guess is that this had changed recently, when some test generators were moved from the kotlin repository to intellij-community.

For what it's worth, I've fixed this one issue by adding a direct dependency on `:compiler:incremental-compilation-impl` to `:generators`. Hopefully this will save a headache for whoever has to eventually migrate the gradle rules. I'm sure there would be many more...

sfs commented:
@max-kammerer Since there were so many TC failures with this PR I've pushed it to prr/sfs-jvm-abi-gen on google/kotlin as well in order to have a look at a TC run from my end. I'll ping you again once the results look okay (with the caveat that I'm never 100% sure which tests are actually failing and which are just temporarily broken).

sfs commented:
Rebased on master and added support for annotation implementation classes. They are handled exactly the same as `WhenMapping` classes (and should probably be generated as package private, but that's a separate issue).

sfs commented:
Extended the test suite to add tests with Java files and language version 1.6 for testing annotation implementation classes. Ideally these tests should use the same infrastructure as the existing codegen tests (and in particular shouldn't use a `LANGUAGE_VERSION` directive), but that will be a bigger refactoring and this PR is already getting rather large.

Of course, after adding tests for `WhenMapping` and annotation implementation classes it turned out that both were incorrectly not stripped even when that was possible. The reason for `WhenMapping` classes was that they are no longer (?) marked as local or anonymous which required a special case in jvm-abi-gen (a1268c5). This is correct, since unlike in Java `WhenMapping` classes are generated as nested synthetic classes in Kotlin. Annotation implementation classes on the other hand are generated as anonymous classes but were not marked as such. This is fixed in e451c2d.

sfs commented on compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/JvmAnnotationImplementationTransformer.kt:
```diff
@@ -32,6 +34,20 @@ internal val annotationImplementationPhase = makeIrFilePhase<JvmBackendContext>(
     AnnotationImplementationTransformer(jvmContext, file) {
+    private val publicAnnotationImplementationClasses = mutableSetOf<IrClassSymbol>()
+
+    // FIXME: Copied from JvmSingleAbstractMethodLowering
```
Before someone asks, the problem is that `allScopes` is protected in `IrElementTransformerVoidWithContext` which is in backend.common while `isInPublicInlineScope` is defined in backend.jvm, i.e., we can't define this utility as an extension function without refactoring somewhere. And I don't want to make such a refactoring part of this PR.

udalov reviewed (approved):
Please update testdata of `IrBytecodeListingTestGenerated.Annotations.testAnnotationCtorCallGenerateSynthetic`.

I've left a few extremely minor comments, but other than that, LGTM.

sfs commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiMetadataProcessor.kt:
```diff
@@ -0,0 +1,194 @@
+    kotlinClassHeaderVisitor { header ->
+        // kotlinx-metadata only supports writing Kotlin metadata of version >= 1.4, so we need to
+        // update the metadata version if we encounter older metadata annotations.
+        val metadataVersion =
```
Done

sfs commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiMetadataProcessor.kt:
```diff
@@ -0,0 +1,194 @@
+
+
+        override fun visitEnd() {
+            body(KotlinClassHeader(
```
Done

sfs commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiMetadataProcessor.kt:
```diff
@@ -0,0 +1,194 @@
+                METADATA_STRINGS_FIELD_NAME -> data2
+                else -> error("Wrong array field in kotlin metadata: $name")
+            }
+            return object : AnnotationVisitor(Opcodes.ASM7) {
```
Fixed

sfs commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiMetadataProcessor.kt:
```diff
@@ -0,0 +1,194 @@
+            val destination = when (name) {
+                METADATA_DATA_FIELD_NAME -> data1
+                METADATA_STRINGS_FIELD_NAME -> data2
+                else -> error("Wrong array field in kotlin metadata: $name")
```
Makes sense. Done

sfs commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiMetadataProcessor.kt:
```diff
@@ -0,0 +1,194 @@
+}
+
+/**
+ * Class metadata adapter which removes private functions, properties, and type aliases.
```
Done

sfs commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiClassBuilderInterceptor.kt:
```diff
@@ -0,0 +1,173 @@
+
+        override fun done() {
+            // Remove local or anonymous classes unless they are in the scope of an inline function and
+            // strip non-inline methods from all other classe.
```
Fixed

sfs commented on plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiOutputExtension.kt:
```diff
@@ -0,0 +1,248 @@
+        val outputFiles = AbiOutputFiles(abiClassInfos, factory)
+        if (outputPath.extension == "jar") {
+            // We don't include the runtime or main class in interface jars and always reset time stamps.
+            CompileEnvironmentUtil.writeToJar(outputPath, false, true, true, null, outputFiles, MessageCollector.NONE)
```
...I completely missed that. Changed the code to use the same `MessageCollector` as the rest of the compiler. That should be fine, since the only diagnostic that `writeToJar` can report is an invalid jar path error.

sfs commented:
Rebased on master and addressed comments in 317caa7 and 614664c.

udalov commented:
Thanks! Please rebase once again and I'll merge it.
(I'd like to merge via GitHub instead of manually, to make GitHub properly display the link to this PR and the discussions here.)

sfs commented:
Rebased on master

udalov commented:
Merged into master, it'll be released in 1.6.0-M1.

@sfs Thank you. Excellent work!
